### PR TITLE
Release Distribution base class and Add Normal, Uniform class

### PR DESCRIPTION
**Pull request by pangyoki**
### PR types

Others

### PR changes

APIs

### Describe

Implement Probability Distribution related classes.
- Release `Distribution` abstract base class and its `sample`, `log_prob`, `entropy`, `probs` methods.
- Move `Normal` class and `Uniform` class in paddle/fluid/layers/distributions.py path to paddle/distribution.py.
- Add parameter `name` of each class API. `name` is initialized in `__init__` function, and every function of the class has its own `name`.
- Add `probs` method for `Normal` class and `Uniform` class.

### CODE REVIEW

File: python/paddle/distribution.py [hunk 1]
```diff
@@ -18,3 +18,457 @@
...
+
+    * :math:`low = a`,
+    * :math:`high = b`,
+    * :math:`Z`: is the normalizing constant.
+
+    The parameters `low` and `high` must be shaped in a way that supports
+    broadcasting (e.g., `high - low` is a valid operation).
+
+    Args:
+        low(float|list|numpy.ndarray|Variable): The lower boundary of uniform distribution.The data type is float32
```
Comment (LutaoChu) [inline python/paddle/distribution.py]: Variable -》 Tensor ，the following is the same

File: python/paddle/distribution.py [hunk 2]
```diff
@@ -18,3 +18,457 @@
...
+        Z = b - a
+
+    In the above equation:
+
+    * :math:`low = a`,
+    * :math:`high = b`,
+    * :math:`Z`: is the normalizing constant.
+
+    The parameters `low` and `high` must be shaped in a way that supports
+    broadcasting (e.g., `high - low` is a valid operation).
```
Comment (LutaoChu) [inline python/paddle/distribution.py]: 加一个broadcast semantic的链接

File: python/paddle/distribution.py [hunk 3]
```diff
@@ -18,3 +18,457 @@
...
+    Args:
+        low(float|list|numpy.ndarray|Variable): The lower boundary of uniform distribution.The data type is float32
+        high(float|list|numpy.ndarray|Variable): The higher boundary of uniform distribution.The data type is float32
+        name(str, optional): Name for the operation (optional, default is None). For more information, please refer to :ref:`api_guide_Name`.
+
+    Examples:
+        .. code-block:: python
+
+          import numpy as np
+          from paddle.fluid import layers
```
Comment (LutaoChu) [inline python/paddle/distribution.py]: 避免出现paddle.fluid

File: python/paddle/distribution.py [hunk 4]
```diff
@@ -18,3 +18,457 @@
...
+          u3 = Uniform(low=[[1.0, 2.0], [3.0, 4.0]],
+                    high=[[1.5, 2.5], [3.5, 4.5]])
+
+          # With broadcasting:
+          u4 = Uniform(low=3.0, high=[5.0, 6.0, 7.0])
+
+          # Complete example
+          value_npdata = np.array([0.8], dtype="float32")
+          value_tensor = layers.create_tensor(dtype="float32")
+          layers.assign(value_npdata, value_tensor)
```
Comment (LutaoChu) [inline python/paddle/distribution.py]: layers接口替换为新版本的paddle接口

File: python/paddle/distribution.py [hunk 5]
```diff
@@ -18,3 +18,457 @@
...
+        batch_shape = list((self.low + self.high).shape)
+        if self.batch_size_unknown:
+            output_shape = shape + batch_shape
+            zero_tmp = tensor.fill_constant_batch_size_like(
+                self.low + self.high, batch_shape + shape, self.low.dtype, 0.)
+            uniform_random_tmp = nn.uniform_random_batch_size_like(
+                zero_tmp, zero_tmp.shape, min=0., max=1., seed=seed)
+            output = uniform_random_tmp * (zero_tmp + self.high - self.low
+                                           ) + self.low
+            return nn.reshape(output, output_shape)
```
Comment (LutaoChu) [inline python/paddle/distribution.py]: return应该要加个name吧，return nn.reshape(output, output_shape, name=name)

File: python/paddle/distribution.py [hunk 6]
```diff
@@ -18,3 +18,457 @@
...
+        else:
+            output_shape = shape + batch_shape
+            output = nn.uniform_random(
+                output_shape, seed=seed) * (tensor.zeros(
+                    output_shape, dtype=self.low.dtype) +
+                                            (self.high - self.low)) + self.low
+            if self.all_arg_is_float:
+                return nn.reshape(output, shape)
```
Comment (LutaoChu) [inline python/paddle/distribution.py]: 同上

File: python/paddle/distribution.py [hunk 7]
```diff
@@ -18,3 +18,457 @@
...
+            else:
+                return output
```
Comment (LutaoChu) [inline python/paddle/distribution.py:259]: 同上，以下各方法的输出都应该有name

File: python/paddle/distribution.py (hunk 1 above)
Comment (LutaoChu) [reply python/paddle/distribution.py]: pytorch文档没写支持int，但实际是支持的。考虑int的情况较常用，建议也支持一下

File: python/paddle/distribution.py [hunk 8]
```diff
@@ -18,3 +18,457 @@
...
+    * :math:`low = a`,
+    * :math:`high = b`,
+    * :math:`Z`: is the normalizing constant.
+
+    The parameters `low` and `high` must be shaped in a way that supports
+    broadcasting (e.g., `high - low` is a valid operation).
+
+    Args:
+        low(float|list|numpy.ndarray|Variable): The lower boundary of uniform distribution.The data type is float32
+        high(float|list|numpy.ndarray|Variable): The higher boundary of uniform distribution.The data type is float32
```
Comment (LutaoChu) [inline python/paddle/distribution.py]: 同上

File: python/paddle/distribution.py [hunk 9]
```diff
@@ -18,3 +18,457 @@
...
+        """Generate samples of the specified shape.
+
+        Args:
+          shape (list): 1D `int32`. Shape of the generated samples.
+          seed (int): Python integer number.
+
+        Returns:
+          Variable: A tensor with prepended dimensions shape.The data type is float32.
+
+        """
```
Comment (LutaoChu) [inline python/paddle/distribution.py:233]: 建议增加动态图的快速路径写法

File: python/paddle/distribution.py [hunk 10]
```diff
@@ -18,3 +18,457 @@
...
+    def log_prob(self, value):
+        """Log probability density/mass function.
+
+        Args:
+          value (Variable): The input tensor.
+
+        Returns:
+          Variable: log probability.The data type is same with value.
+
+        """
```
Comment (LutaoChu) [inline python/paddle/distribution.py:270]: 建议增加动态图快速路径写法，以下所有方法同样建议增加

File: python/paddle/distribution.py [hunk 11]
```diff
@@ -18,3 +18,457 @@
...
+
+        Args:
+          value (Variable): The input tensor.
+
+        Returns:
+          Variable: probability.The data type is same with value.
+
+        """
+        name = self.name + '_probs'
+        return ops.exp(self.log_prob(value))
```
Comment (LutaoChu) [inline python/paddle/distribution.py]: 建议直接用概率定义计算

File: python/paddle/distribution.py [hunk 12]
```diff
@@ -18,3 +18,457 @@
...
+        Z = (2 \pi \sigma^2)^{0.5}
+
+    In the above equation:
+
+    * :math:`loc = \mu`: is the mean.
+    * :math:`scale = \sigma`: is the std.
+    * :math:`Z`: is the normalization constant.
+
+    Args:
+        loc(float|list|numpy.ndarray|Variable): The mean of normal distribution.The data type is float32.
```
Comment (LutaoChu) [inline python/paddle/distribution.py]: 同Uniform，Variable -》Tensor， 建议支持int

File: python/paddle/distribution.py [hunk 13]
```diff
@@ -18,3 +18,457 @@
...
+          dist.sample([3])
+
+          # Define a batch of two scalar valued Normals.
+          # Both have mean 1, but different standard deviations.
+          dist = Normal(loc=1., scale=[11., 22.])
+
+          # Complete example
+          value_npdata = np.array([0.8], dtype="float32")
+          value_tensor = layers.create_tensor(dtype="float32")
+          layers.assign(value_npdata, value_tensor)
```
Comment (LutaoChu) [inline python/paddle/distribution.py]: value_tensor建议用paddle.to_tensor创建

File: python/paddle/distribution.py [hunk 14]
```diff
@@ -18,3 +18,457 @@
...
+
+        Args:
+          value (Variable): The input tensor.
+
+        Returns:
+          Variable: probability.The data type is same with value.
+
+        """
+        name = self.name + '_probs'
+        return ops.exp(self.log_prob(value))
```
Comment (LutaoChu) [inline python/paddle/distribution.py]: 同Uniform，建议直接计算pdf

File: python/paddle/fluid/tests/unittests/test_distribution.py [hunk 15]
```diff
@@ -0,0 +1,542 @@
...
+        normal2 = Normal(0.0, 1.0)
+        self.assertEqual(normal2.name, 'Normal')
+
+    def test_uniform_name(self):
+        name = 'test_uniform'
+        uniform1 = Uniform(0.0, 1.0, name=name)
+        self.assertEqual(uniform1.name, name)
+
+        uniform2 = Uniform(0.0, 1.0)
+        self.assertEqual(uniform2.name, 'Uniform')
```
Comment (LutaoChu) [inline python/paddle/fluid/tests/unittests/test_distribution.py:703]: 补充method name的单测用例

File: python/paddle/distribution.py (hunk 1 above)
Comment (pangyoki) [reply python/paddle/distribution.py]: Done

File: python/paddle/distribution.py (hunk 2 above)
Comment (pangyoki) [reply python/paddle/distribution.py]: Done

File: python/paddle/distribution.py (hunk 3 above)
Comment (pangyoki) [reply python/paddle/distribution.py]: Done

File: python/paddle/distribution.py (hunk 4 above)
Comment (pangyoki) [reply python/paddle/distribution.py]: Done, use `paddle.to_tensor` method

File: python/paddle/distribution.py (hunk 5 above)
Comment (pangyoki) [reply python/paddle/distribution.py]: Done

File: python/paddle/distribution.py (hunk 11 above)
Comment (pangyoki) [reply python/paddle/distribution.py]: Done

File: python/paddle/distribution.py (hunk 12 above)
Comment (pangyoki) [reply python/paddle/distribution.py]: Done

File: python/paddle/distribution.py (hunk 13 above)
Comment (pangyoki) [reply python/paddle/distribution.py]: Done

File: python/paddle/distribution.py (hunk 14 above)
Comment (pangyoki) [reply python/paddle/distribution.py]: Done

File: python/paddle/fluid/tests/unittests/test_distribution.py (hunk 15 above)
Comment (pangyoki) [reply python/paddle/fluid/tests/unittests/test_distribution.py:703]: Done

File: python/paddle/distribution.py (hunk 6 above)
Comment (pangyoki) [reply python/paddle/distribution.py]: Done

File: python/paddle/distribution.py (hunk 7 above)
Comment (pangyoki) [reply python/paddle/distribution.py:259]: Done

File: python/paddle/distribution.py (hunk 8 above)
Comment (pangyoki) [reply python/paddle/distribution.py]: Done

File: python/paddle/distribution.py (hunk 9 above)
Comment (pangyoki) [reply python/paddle/distribution.py:233]: Done

File: python/paddle/distribution.py (hunk 10 above)
Comment (pangyoki) [reply python/paddle/distribution.py:270]: Done

File: python/paddle/distribution.py [hunk 16]
```diff
@@ -18,3 +18,517 @@
...
+
+    * :math:`low = a`,
+    * :math:`high = b`,
+    * :math:`Z`: is the normalizing constant.
+
+    The parameters `low` and `high` must be shaped in a way that supports
+    [broadcasting](https://www.paddlepaddle.org.cn/documentation/docs/en/develop/beginners_guide/basic_concept/broadcasting_en.html) (e.g., `high - low` is a valid operation).
+
+    Args:
+        low(int|float|list|numpy.ndarray|Tensor): The lower boundary of uniform distribution.The data type is float32 or int
```
Comment (LutaoChu) [inline python/paddle/distribution.py:157]: 漏了list|numpy.ndarray|Tensor

File: python/paddle/distribution.py [hunk 17]
```diff
@@ -18,3 +18,517 @@
...
+        high(int|float|list|numpy.ndarray|Tensor): The higher boundary of uniform distribution.The data type is float32 or int
```
Comment (LutaoChu) [inline python/paddle/distribution.py:158]: 同上

File: python/paddle/distribution.py [hunk 18]
```diff
@@ -18,3 +18,517 @@
...
+
+        """
+        name = self.name + '_log_prob'
+        if in_dygraph_mode():
+            lb_bool = self.low < value
+            ub_bool = value < self.high
+            lb = tensor.cast(lb_bool, dtype=value.dtype)
+            ub = tensor.cast(ub_bool, dtype=value.dtype)
+            return elementwise_sub(
+                nn.log(lb * ub), nn.log(self.high - self.low), name=name)
```
Comment (LutaoChu) [inline python/paddle/distribution.py:278]: 动态图不需要加name，cast和elementwise_sub最好用core.ops

File: python/paddle/distribution.py [hunk 19]
```diff
@@ -18,3 +18,517 @@
...
+          Tensor: probability.The data type is same with value.
+
+        """
+        name = self.name + '_probs'
+        if in_dygraph_mode():
+            lb_bool = self.low < value
+            ub_bool = value < self.high
+            lb = tensor.cast(lb_bool, dtype=value.dtype)
+            ub = tensor.cast(ub_bool, dtype=value.dtype)
+            return elementwise_div((lb * ub), (self.high - self.low), name=name)
```
Comment (LutaoChu) [inline python/paddle/distribution.py:306]: 同上

File: python/paddle/distribution.py [hunk 20]
```diff
@@ -18,3 +18,517 @@
...
+        name = self.name + '_entropy'
+        return nn.log(self.high - self.low, name=name)
+
+
+class Normal(Distribution):
+    """The Normal distribution with location `loc` and `scale` parameters.
+
+    Mathematical details
+
+    The probability density function (pdf) is,
```
Comment (LutaoChu) [inline python/paddle/distribution.py:333]: 去掉逗号

File: python/paddle/distribution.py [hunk 21]
```diff
@@ -18,3 +18,517 @@
...
+
+    In the above equation:
+
+    * :math:`loc = \mu`: is the mean.
+    * :math:`scale = \sigma`: is the std.
+    * :math:`Z`: is the normalization constant.
+
+    Args:
+        loc(int|float|list|numpy.ndarray|Tensor): The mean of normal distribution.The data type is float32 or int.
+        scale(int|float|list|numpy.ndarray|Tensor): The std of normal distribution.The data type is float32 or int.
```
Comment (LutaoChu) [inline python/paddle/distribution.py:351]: data type缺少list|numpy.ndarray|Tensor

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Review by raindrops2sea — APPROVED** [review]

**Review by zhiqiu — APPROVED** [review]

**Review by LutaoChu — APPROVED** [review]
LGTM

**Review by saxon-zh — APPROVED** [review]
LGTM

File: python/paddle/distribution.py (hunk 16 above)
Comment (pangyoki) [reply python/paddle/distribution.py:157]: Done, refer to PR #26535

File: python/paddle/distribution.py (hunk 17 above)
Comment (pangyoki) [reply python/paddle/distribution.py:158]: Done, refer to PR #26535

File: python/paddle/distribution.py (hunk 18 above)
Comment (pangyoki) [reply python/paddle/distribution.py:278]: Done, refer to PR #26535

File: python/paddle/distribution.py (hunk 19 above)
Comment (pangyoki) [reply python/paddle/distribution.py:306]: Done, refer to PR #26535

File: python/paddle/distribution.py (hunk 20 above)
Comment (pangyoki) [reply python/paddle/distribution.py:333]: Done, refer to PR #26535

File: python/paddle/distribution.py (hunk 21 above)
Comment (pangyoki) [reply python/paddle/distribution.py:351]: Done, refer to PR #26535

**Pull request merged**